File: Software/MenuPrototype/OptionChooser.cs
using System;
using System.Linq;

namespace MenuPrototype
{
	partial class Program
	{
		/*
		ModalScreen - Takes over display
		ActionList - Select from list of items to invoke an action(used in GO)
		LabelList - Scroll through list of items(used in INFO)
		SubmenuList - Choose items(CTRL / CAL)
		ModalSubmenuList - Only for SubMenu, takes over 2nd line
		*/

		////////////////////////////////////////////////////////////////////
		public class OptionChooser : MenuItem
		{ // 2 (Yes/No) or 3 (Yes/No/Cancl)
			private int _activeOption;
			string[] _choices;
			Action<EventArgs> _chosenFunc;
			public OptionChooser(string[] choices, int activeItemIndex, Action<EventArgs> chosenFunc) : base("")
			{
				_activeOption = activeItemIndex;
				_chosenFunc = chosenFunc;
				_choices = choices.ToArray();
			}

			public override void onDisplay(bool modal = false)
			{
				string line = string.Empty;
				for (int i = 0; i < _choices.Length; i++)
				{
					line += ((i == _activeOption) ? '>' : ' ') + _choices[i] + ((i == _activeOption) ? '<' : ' ') + ' ';
				}
				Console.WriteLine(line);
			}

			public override void onSelect()
			{
				_chosenFunc(new SelectEventArgs(this, _choices[_activeOption]));
			}

			public string getSelected()
			{
				return _choices[_activeOption];
			}

			public override bool onKeypressed(int key)
			{
				if (base.onKeypressed(key)) { return true; }
				if (key == btnLEFT)
				{
					_activeOption = (_activeOption + 1) % _choices.Length;
					return true;
				}
				return false;
			}
		}
	}
}

File: Software/Arduino code/OpenAstroTracker/lib/menu/controls/OptionChooser.cpp

#include "MainMenu.hpp"
#include "SelectEventArgs.hpp"
#include "OptionChooser.hpp"
#include "../../input/LcdButtons.hpp"

OptionChooser::OptionChooser(String choice1, String choice2, int activeItemIndex, eventAction chosenFunc) : MenuItem("", "OPT")
{
    _activeOption = activeItemIndex;
    _chosenFunc = chosenFunc;
    _choices[0] = choice1;
    _choices[1] = choice2;
    _numChoices = 2;
}

OptionChooser::OptionChooser(String choice1, String choice2, String choice3, int activeItemIndex, eventAction chosenFunc) : MenuItem("", "OPT")
{
    _activeOption = activeItemIndex;
    _chosenFunc = chosenFunc;
    _choices[0] = choice1;
    _choices[1] = choice2;
    _choices[2] = choice3;
    _numChoices = 3;
}

OptionChooser::OptionChooser(String choice1, String choice2, String choice3, String choice4, int activeItemIndex, eventAction chosenFunc) : MenuItem("", "OPT")
{
    _activeOption = activeItemIndex;
    _chosenFunc = chosenFunc;
    _choices[0] = choice1;
    _choices[1] = choice2;
    _choices[2] = choice3;
    _choices[3] = choice4;
    _numChoices = 4;
}

void OptionChooser::onDisplay(bool modal)
{
    String line = "";
    for (int i = 0; i < _numChoices; i++)
    {
        line += ((i == _activeOption) ? '>' : ' ') + _choices[i] + ((i == _activeOption) ? '<' : ' ') + ' ';
    }
    if (line.length() > 16)
    {
        line = "";
        for (int i = 0; i < _numChoices; i++)
        {
            line += ((i == _activeOption) ? '>' : ' ') + _choices[i] + ((i == _activeOption) ? '<' : ' ');
        }
    }

    getMainMenu()->writeToLCD(0, 1, line);
}

void OptionChooser::onSelect()
{
    SelectEventArgs chosen(this, _choices[_activeOption]);
    _chosenFunc(&chosen);
}

String OptionChooser::getSelected()
{
    return _choices[_activeOption];
}

int OptionChooser::getSelectedIndex()
{
    return _activeOption;
}

bool OptionChooser::onKeypressed(int key)
{
    if (MenuItem::onKeypressed(key))
    {
        return true;
    }

    if (key == btnLEFT)
    {
        _activeOption = (_activeOption + 1) % _numChoices;
        return true;
    }
    return false;
}
